src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt

```diff
@@ -415,8 +426,41 @@ class CosmasGoogleCloudService(private val bucketName: String,
             handleStorageException(e, responseObserver)
             return
         }
-        val response = CosmasProto.RestoreDeletedFileResponse.newBuilder()
-        responseObserver.onNext(response.build())
+        val response = CosmasProto.RestoreDeletedFileResponse.getDefaultInstance()
+        responseObserver.onNext(response)
+        responseObserver.onCompleted()
+    }
+
+    override fun changeFileId(request: CosmasProto.ChangeFileIdRequest, responseObserver: StreamObserver<CosmasProto.ChangeFileIdResponse>) {
+        LOG.info("Get request for change files ids in project # ${request.projectId}")
+        val prevIds = try {
+            getPrevIds(request.projectId).toMutableMap()
+        } catch (e: StorageException) {
+            handleStorageException(e, responseObserver)
+            return
+        }
+        val project = synchronized(this.fileBuffer) {
+            this.fileBuffer.getValue(request.projectId)
+        }
+        for (change in request.changesList) {
+            prevIds[change.newFileId] = change.oldFileId
+            synchronized(project) {
+                if (project.contains(change.oldFileId)) {
+                    project[change.newFileId] = project[change.oldFileId]
+                    project.remove(change.oldFileId)
+                }
+            }
+        }
+        try {
```

Comment:
I suggest making it more reliable: commit project on receiving ChangeFileIdRequest ( => you saved the history associated with all old ids), clear `project` ( => you do not need to remap ids, and new ids will start filling `project` from scratch) and make sure you write the mapping. The order of project commit/write map events doesn't seem to be important, but if you do write map first and then commit, holding the lock on `project` during all that process, it will be more reliable.